README.md
# leetcode-cli

![Rust](https://github.com/clearloop/leetcode-cli/workflows/leetcode-cli/badge.svg)
[![crate](https://img.shields.io/crates/v/leetcode-cli.svg)](https://crates.io/crates/leetcode-cli)
[![doc](https://img.shields.io/badge/current-docs-brightgreen.svg)](https://docs.rs/leetcode-cli/)
[![downloads](https://img.shields.io/crates/d/leetcode-cli.svg)](https://crates.io/crates/leetcode-cli)
[![telegram](https://img.shields.io/badge/telegram-blue?logo=telegram)](https://t.me/+U_5si6PhWykxZTI1)
[![LICENSE](https://img.shields.io/crates/l/leetcode-cli.svg)](https://choosealicense.com/licenses/mit/)

## Installing

```sh
# Required dependencies:
#
#  gcc
#  libssl-dev
#  libdbus-1-dev
#  libsqlite3-dev

cargo install leetcode-cli
```

## Usage

**Make sure you have logged in to `leetcode.com` with `Chrome`**. See [Cookies](#cookies) for why you need to do this first.

```sh
leetcode 0.4.0
May the Code be with You 👻

USAGE:
    leetcode [FLAGS] [SUBCOMMAND]

FLAGS:
    -d, --debug      debug mode
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    data    Manage Cache [aliases: d]
    edit    Edit question by id [aliases: e]
    exec    Submit solution [aliases: x]
    list    List problems [aliases: l]
    pick    Pick a problem [aliases: p]
    stat    Show simple chart about submissions [aliases: s]
    test    Edit question by id [aliases: t]
    help    Prints this message or the help of the given subcommand(s)
```

## Example

For example, given this config (could be found at `~/.leetcode/leetcode.toml`):

```toml
[code]
editor = 'emacs'
# Optional parameter
editor-args = ['-nw']
lang = 'rust'

[cookies]
csrf = '<your-leetcode-csrf-token>'
session = '<your-leetcode-session-key>'

[storage]
cache = 'Problems'
code = 'code'
root = '~/.leetcode'
scripts = 'scripts'
```

#### 1. <kbd>pick</kbd>

```sh
leetcode pick 1
```

```sh
[1] Two Sum is on the run...


Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

--------------------------------------------------

Example:


Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
```

#### 2. <kbd>edit</kbd>

```sh
leetcode edit 1
```

```rust
# struct Solution;
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        use std::collections::HashMap;
        let mut m: HashMap<i32, i32> = HashMap::new();

        for (i, e) in nums.iter().enumerate() {
            if let Some(v) = m.get(&(target - e)) {
                return vec![*v, i as i32];
            }

            m.insert(*e, i as i32).unwrap_or_default();
        }

        return vec![];
    }
}
```

#### 3. <kbd>test</kbd>

```sh
leetcode test 1
```

```sh

  Accepted       Runtime: 0 ms

  Your input:    [2,7,11,15], 9
  Output:        [0,1]
  Expected:      [0,1]

```

#### 4. <kbd>exec</kbd>

```sh
leetcode exec 1
```

```sh

  Success

  Runtime: 0 ms, faster than 100% of Rustonline submissions for Two Sum.

  Memory Usage: 2.4 MB, less than 100% of Rustonline submissions for Two Sum.


```

## Cookies

The cookie plugin of leetcode-cli can work on OSX and [Linux][#1]. **If you are on a different platform, there are problems with caching the cookies**,
you can manually input your LeetCode Cookies to the configuration file.

```toml
[cookies]
csrf = "..."
session = "..."
```

For Example, using Chrome (after logging in to LeetCode):

#### Step 1

Open Chrome and navigate to the link below:

```sh
chrome://settings/cookies/detail?site=leetcode.com
```

#### Step 2

Copy `Content` from `LEETCODE_SESSION` and `csrftoken` to `session` and `csrf` in your configuration file, respectively:

```toml
[cookies]
csrf = "${csrftoken}"
session = "${LEETCODE_SESSION}"
```

## Programmable

If you want to filter LeetCode questions using custom Python scripts, add the following to your the configuration file:

```toml
[storage]
scripts = "scripts"
```

Then write the script:

```python
# ~/.leetcode/scripts/plan1.py
import json;

def plan(sps, stags):
    ##
    # `print` in python is supported,
    # if you want to know the data structures of these two args,
    # just print them
    ##
    problems = json.loads(sps)
    tags = json.loads(stags)

    ret = []
    tm = {}
    for tag in tags:
        tm[tag["tag"]] = tag["refs"];

    for i in problems:
        if i["level"] == 1 and str(i["id"]) in tm["linked-list"]:
            ret.append(str(i["id"]))

    # return is `List[string]`
    return ret
```

Then run `list` with the filter that you just wrote:

```sh
leetcode list -p plan1
```

That's it! Enjoy!

## Contributions

Feel free to add your names and emails in the `authors` field of `Cargo.toml` !

## LICENSE

MIT

[pr]: https://github.com/clearloop/leetcode-cli/pulls
[#1]: https://github.com/clearloop/leetcode-cli/issues/1


CHANGELOG.md
## v0.4.1

- Search problems by name
- Re-enable chrome plugin

## v0.3.3

- allow more flexible categories by @frrad
- change params type to `Option<Param>` @frrad

## v0.3.2

- adds additional tracing by @shmuga
- removes test_mode parameter by @shmuga

## v0.3.1

- pipe handling by @aymanbagabas
- Improve README by @xiaoxiae

## v0.3.0

- Upgrade reqwest to async mode
- Format code using clippy

## v0.2.23

- support color display

## v0.2.22

- Fixed the cache can't update with new added problems

- Display user friendly errors when pick/edit new added problem.

- upgrade pyo3

- fix leetcode list with empty cache

## v0.2.21

- Make programmable support to be an advanced feature

## v0.2.20

- Support sup/sub style for numbers

## v0.2.19

- Better HTML!

## v0.2.18

- Display stdout for test and execute commands, fix minor spacing in results displayed

- Fix panic on `pick` command without cache

## v0.2.17

Fix panic on stat command with zero numbers

## v0.2.16

Update versions of diesel and reqwest

## v0.2.15

Allow for custom testcases with the `leetcode test` command, and some minor edits

## v0.2.14

Corrects file suffixes for c\*\* and c# files

## v0.2.13

fix percent length panic

## v0.2.12

fix gt || ge || lt || le

## v0.2.11

added code 14 and transfered `&#ge;`、`&#le` and `&#39;`.

## v0.2.10

add code 15

## v0.2.9

update ac status after submit successfully

## v0.2.8

show last testcases

## v0.2.7

fixed float bug in result

## v0.2.6

sync config while change current lang

## v0.2.5

update local cache when submission status changes

## v0.2.4

auto fetch question while exec `edit` directly.

## v0.2.3

Programmable leetcode-cli

## v0.2.2

1. optimize logs
2. add tag filter
3. sync configs

## v0.2.1

1. fix cookies error handling
2. dismiss all `unwrap`, `expect` and `panic`
3. add cookie configs

## v0.2.0

1. Add Linux Support

## v0.1.9

1. release submit command
2. deserialize json using outter funcs

## v0.1.8

1. pack mod exports
2. add edit command
3. add test command

## v0.1.7

render html in command-line, and `pick` command

## v0.1.6

complete `stat` command

## v0.1.5

complete `cache` command

## v0.1.3

complete `list` command

## v0.1.2

abstract data cache

## v0.1.1

add list command

## v0.1.0

chrome cookie


src/lib.rs
//! # leetcode-cli
//! [![doc](https://img.shields.io/badge/current-docs-green.svg)](https://docs.rs/leetcode-cli/)
//! [![Crates.io](https://img.shields.io/crates/v/leetcode-cli.svg)](https://crates.io/crates/leetcode-cli)
//! [![Crates.io](https://img.shields.io/crates/d/leetcode-cli.svg)](https://crates.io/crates/leetcode-cli)
//! [![LICENSE](https://img.shields.io/crates/l/leetcode-cli.svg)](https://choosealicense.com/licenses/mit/)
//!
//! ## Installing
//!
//! ```sh
//! cargo install leetcode-cli
//! ```
//!
//! ## Usage
//!
//! **Please make sure you have logined in `leetcode.com` with `chrome`**, more info plz checkout [this](#cookies)
//!
//! ```sh
//! leetcode 0.3.10
//! May the Code be with You 👻
//!
//! USAGE:
//!     leetcode [FLAGS] [SUBCOMMAND]
//!
//! FLAGS:
//!     -d, --debug      debug mode
//!     -h, --help       Prints help information
//!     -V, --version    Prints version information
//!
//! SUBCOMMANDS:
//!     data    Manage Cache [aliases: d]
//!     edit    Edit question by id [aliases: e]
//!     exec    Submit solution [aliases: x]
//!     list    List problems [aliases: l]
//!     pick    Pick a problem [aliases: p]
//!     stat    Show simple chart about submissions [aliases: s]
//!     test    Edit question by id [aliases: t]
//!     help    Prints this message or the help of the given subcommand(s)
//! ```
//!
//! ## Example
//!
//! For example, if your config is:
//!
//! ```toml
//! [code]
//! lang = "rust"
//! editor = "emacs"
//! ```
//!
//! #### 1. <kbd>pick</kbd>
//!
//! ```sh
//! leetcode pick 1
//! ```
//!
//! ```sh
//! [1] Two Sum is on the run...
//!
//!
//! Given an array of integers, return indices of the two numbers such that they add up to a specific target.
//!
//! You may assume that each input would have exactly one solution, and you may not use the same element twice.
//!
//! --------------------------------------------------
//!
//! Example:
//!
//!
//! Given nums = [2, 7, 11, 15], target = 9,
//!
//! Because nums[0] + nums[1] = 2 + 7 = 9,
//! return [0, 1].
//! ```
//!
//! #### 2. <kbd>edit</kbd>
//!
//! ```sh
//! leetcode edit 1
//! ```
//!
//! ```rust
//! # struct Solution;
//! impl Solution {
//!     pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
//!         use std::collections::HashMap;
//!         let mut m: HashMap<i32, i32> = HashMap::new();
//!
//!         for (i, e) in nums.iter().enumerate() {
//!             if let Some(v) = m.get(&(target - e)) {
//!                 return vec![*v, i as i32];
//!             }
//!
//!             m.insert(*e, i as i32).unwrap_or_default();
//!         }
//!
//!         return vec![];
//!     }
//! }
//! ```
//!
//! #### 3. <kbd>test</kbd>
//!
//! ```sh
//! leetcode test 1
//! ```
//!
//! ```sh
//!
//!   Accepted       Runtime: 0 ms
//!
//!   Your input:    [2,7,11,15], 9
//!   Output:        [0,1]
//!   Expected:      [0,1]
//!
//! ```
//!
//! #### 4. <kbd>submit</kbd>
//!
//! ```sh
//! leetcode submit 1
//! ```
//!
//! ```sh
//!
//!   Success
//!
//!   Runtime: 0 ms, faster than 100% of Rustonline submissions for Two Sum.
//!
//!   Memory Usage: 2.4 MB, less than 100% of Rustonline submissions for Two Sum.
//!
//!
//! ```
//!
//! ## Cookies
//!
//! The cookie plugin of leetcode-cil can work on OSX and [Linux][#1], **If you are on other platforms or your cookies just don't want to be catched**, you can handwrite your LeetCode Cookies to `~/.leetcode/leetcode.toml`
//!
//! ```toml
//! # Make sure `leetcode.toml` file is placed at `~/.leetcode/leetcode.toml`
//! [cookies]
//! csrf = "..."
//! session = "..."
//! ```
//!
//! For Example, if you're using chrome to login to leetcode.com.
//!
//!
//! #### Step 1
//!
//! Open chrome and paste the link below to the `chrome linkbar`.
//!
//! ```sh
//! chrome://settings/cookies/detail?site=leetcode.com
//! ```
//!
//! #### Step 2
//!
//! Copy the contents of `LEETCODE_SESSION` and `csrftoken`.
//!
//! #### Step 3
//!
//! Paste them to `session` and `csrf`.
//!
//! ```toml
//! # Make sure `leetcode.toml` file is placed at `~/.leetcode/leetcode.toml`
//! [cookies]
//! csrf = "${csrftoken}"
//! session = "${LEETCODE_SESSION}"
//! ```
//!
//!
//! ## Programmable
//!
//! If we want to filter leetcode questions using our own python scripts, what should we do?
//!
//! For example, our config is:
//!
//! ```toml
//! # Make sure `leetcode.toml` file is placed at `~/.leetcode/leetcode.toml`
//! [storage]
//! scripts = "scripts"
//! ```
//!
//! We write our python scripts:
//!
//! ```python
//! # ~/.leetcode/scripts/plan1.py
//! import json;
//!
//! def plan(sps, stags):
//!     ##
//!     # `print` in python is supported,
//!     # if you want to know the data structures of these two args,
//!     # just print them
//!     ##
//!     problems = json.loads(sps)
//!     tags = json.loads(stags)
//!
//!     ret = []
//!     tm = {}
//!     for tag in tags:
//!         tm[tag["tag"]] = tag["refs"];
//!
//!     for i in problems:
//!         if i["level"] == 1 and str(i["id"]) in tm["linked-list"]:
//!             ret.append(str(i["id"]))
//!
//!     # return is `List[string]`
//!     return ret
//! ```
//!
//! Then we can run filter as what we write now:
//!
//! ```sh
//! leetcode list -p plan1
//! ```
//!
//! Well done, enjoy it!
//!
//!
//! ## PR
//!
//! PR is welcome, [here][pr] it is.
//!
//! ## LICENSE
//! MIT
//!
//!
//! [pr]: https://github.com/clearloop/leetcode-cli/pulls
//! [#1]: https://github.com/clearloop/leetcode-cli/issues/1
#[macro_use]
extern crate log;
#[macro_use]
extern crate diesel;

// show docs
pub mod cache;
pub mod cli;
pub mod cmds;
pub mod config;
pub mod err;
pub mod flag;
pub mod helper;
pub mod plugins;
#[cfg(feature = "pym")]
pub mod pym;

// re-exports
pub use cache::Cache;
pub use config::Config;
pub use err::Error;
